src/System.Management.Automation/engine/parser/TypeResolver.cs

```diff
@@ -302,10 +321,28 @@ internal static Type ResolveTypeNameWithContext(TypeName typeName, out Exception
                 return typeName._typeDefinitionAst.Type;
             }
 
-            result = ResolveTypeNameWorker(typeName, currentScope, typeResolutionState.assemblies, typeResolutionState, /* reportAmbiguousException */ true, out exception);
+            var assemList = assemblies ?? ClrFacade.GetAssemblies(typeResolutionState, typeName);
+            // A set of assembly names that we have searched from but found no matching type.
+            // This set will be updated within ResolveTypeNameWorker.
+            var searchedAssemblies = new HashSet<string>();
+
+            result = ResolveTypeNameWorker(typeName, currentScope, typeResolutionState.assemblies, searchedAssemblies, typeResolutionState,
+                                           /*onlySearchInGivenAssemblies*/ false, /* reportAmbiguousException */ true, out exception);
             if (exception == null && result == null)
             {
-                result = ResolveTypeNameWorker(typeName, currentScope, assemList, typeResolutionState, /* reportAmbiguousException */ false, out exception);
+                if (context != null && assemblies == null)
+                {
+                    // If the assemblies to search from is not specified by the caller, then we search our assembly cache first,
+                    // so as to give preference to resolving the type against assemblies explicitly loaded by powershell.
+                    result = ResolveTypeNameWorker(typeName, currentScope, context.AssemblyCache.Values, searchedAssemblies, typeResolutionState,
+                                                   /*onlySearchInGivenAssemblies*/ true, /* reportAmbiguousException */ false, out exception);
+                }
+
+                if (result == null)
+                {
+                    result = ResolveTypeNameWorker(typeName, currentScope, assemList, searchedAssemblies, typeResolutionState,
```

Comment:
Do we detect any ambiguities this way?

There is a balance between successfully finding something, and sometimes finding the wrong thing.

With this approach, we can't detect sometimes finding the wrong thing - which detecting an ambiguity can.